flink-end-to-end-tests/flink-stream-stateful-job-upgrade-test/src/main/java/org/apache/flink/streaming/tests/StatefulStreamJobUpgradeTestProgram.java

```
@@ -0,0 +1,128 @@
+ * The gaol is to test successful state restoration after taking savepoint and recovery with new job version.
+ * It can be configured with '--test.job.variant' to run different variants of it:
+ * <ul>
+ *     <li><b>original:</b> includes 2 custom stateful map operators</li>
+ *     <li><b>upgraded:</b> changes order of 2 custom stateful map operators and adds one more</li>
+ * </ul>
+ */
+public class StatefulStreamJobUpgradeTestProgram {
+	private static final String TEST_JOB_VARIANT_ORIGINAL = "original";
+	private static final String TEST_JOB_VARIANT_UPGRADED = "upgraded";
+
+	private static final JoinFunction<Event, ComplexPayload, ComplexPayload> SIMPLE_STATE_UPDATE =
+		(Event first, ComplexPayload second) -> new ComplexPayload(first);
+	private static final JoinFunction<Event, ComplexPayload, ComplexPayload> LAST_EVENT_STATE_UPDATE =
+		(Event first, ComplexPayload second) ->
+			(second != null && first.getEventTime() <= second.getEventTime()) ? second : new ComplexPayload(first);
+
+	private static final ConfigOption<String> TEST_JOB_VARIANT = ConfigOptions
+		.key("test.job.variant")
+		.defaultValue(TEST_JOB_VARIANT_ORIGINAL)
+		.withDescription(String.format("This configures the job variant to test. Can be '%s' or '%s'",
+			TEST_JOB_VARIANT_ORIGINAL, TEST_JOB_VARIANT_UPGRADED));
+
+	public static void main(String[] args) throws Exception {
+		final ParameterTool pt = ParameterTool.fromArgs(args);
+
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		setupEnvironment(env, pt);
+
+		KeyedStream<Event, Integer> source = env.addSource(createEventSource(pt))
+			.assignTimestampsAndWatermarks(createTimestampExtractor(pt))
+			.keyBy(Event::getKey);
+
+		List<TypeSerializer<ComplexPayload>> stateSer =
+			Collections.singletonList(new KryoSerializer<>(ComplexPayload.class, env.getConfig()));
+
+		boolean isOriginal = pt.get(TEST_JOB_VARIANT.key()).equals(TEST_JOB_VARIANT_ORIGINAL);
```

We could throw an `IllegalArgumentException` for any unexpected string.